KEYCLOAK-9536 DB Allocator Maven Plugin

slaskawi opened the pull request:
https://issues.jboss.org/browse/KEYCLOAK-9536

Here are some use case examples:

A successful DB Allocation (note, I replaced some of the URLs):
```
mvn org.keycloak.testsuite:db-allocator-plugin:6.0.0-SNAPSHOT:allocate -Ddballocator.uri="http://dballocator.keycloak.org:8080/Allocator/AllocatorServlet" -Ddballocator.user=keycloak -Ddballocator.type=mysql57 -Ddballocator.expirationMin=10 -Ddballocator.retry.totalRetries=3

[INFO] -------------< org.keycloak.testsuite:db-allocator-plugin >-------------
[INFO] Building DB Allocator Plugin 6.0.0-SNAPSHOT
[INFO] ----------------------------[ maven-plugin ]----------------------------
[INFO] 
[INFO] --- db-allocator-plugin:6.0.0-SNAPSHOT:allocate (default-cli) @ db-allocator-plugin ---
[info] Allocated database:
[info] -- UUID: 7a55eeeb-162e-465c-83ec-cac61143a09f
[info] -- Driver: com.mysql.jdbc.Driver
[info] -- Database: dballo04
[info] -- User: dballo04
[info] -- Password: dballo04
[info] -- URL: jdbc:mysql://mysql-57.keycloak.org:3306/dballo04
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.813 s
[INFO] Finished at: 2019-03-19T12:35:13+01:00
[INFO] ------------------------------------------------------------------------
```

An unsuccessful DB Allocation:

```
[INFO] -------------< org.keycloak.testsuite:db-allocator-plugin >-------------
[INFO] Building DB Allocator Plugin 6.0.0-SNAPSHOT
[INFO] ----------------------------[ maven-plugin ]----------------------------
[INFO] 
[INFO] --- db-allocator-plugin:6.0.0-SNAPSHOT:allocate (default-cli) @ db-allocator-plugin ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.490 s
[INFO] Finished at: 2019-03-19T12:33:19+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.keycloak.testsuite:db-allocator-plugin:6.0.0-SNAPSHOT:allocate (default-cli) on project db-allocator-plugin: An error occurred while communicating with DBAllocator ([500](
[ERROR] <html>
[ERROR]   <head>
[ERROR]     <title>DBAllocator Error Page</title>
[ERROR]   </head>
[ERROR]   <body style="font-family:monospace">
[ERROR]     <h3>Error in DBAllocator</h3>
[ERROR]     <table>
[ERROR]       <col width="15%">
[ERROR]       <col width="85%">
[ERROR]       <tr>
[ERROR]         <td><strong>Http Status:</strong></td>
[ERROR]         <td>500</td>
[ERROR]       </tr>
[ERROR]       <tr>
[ERROR]         <td><strong>Exception class:</strong></td>
[ERROR]         <td>javax.servlet.ServletException</td>
[ERROR]       </tr>
[ERROR]       <tr>
[ERROR]          <td><strong>Message:</strong></td>
[ERROR]          <td><pre>Allocation failed: org.jboss.qa.dballoc.allocator.impl.exception.ResourceNotAvailableException:No resources found for allocation with label expression: mariadb_galera_101&&geo_RDU
[ERROR] org.jboss.qa.dballoc.allocator.impl.exception.ResourceNotAvailableException: No resources found for allocation with label expression: mariadb_galera_101&&geo_RDU
[ERROR] 	at org.jboss.qa.dballoc.allocator.impl.service.AllocatorServiceImpl.allocate(AllocatorServiceImpl.java:108)
[ERROR] 	at org.jboss.qa.dballoc.allocator.impl.service.AllocatorServiceImpl$Proxy$_$$_WeldClientProxy.allocate(AllocatorServiceImpl$Proxy$_$$_WeldClientProxy.java)
[ERROR] 	at org.jboss.qa.dballoc.allocator.impl.rest.DbAllocatorResourceImpl.actuallyAllocate(DbAllocatorResourceImpl.java:23)
[ERROR] 	at org.jboss.qa.dballoc.allocator.impl.rest.DbAllocatorResourceImpl.allocate(DbAllocatorResourceImpl.java:36)
[ERROR] 	at org.jboss.qa.dballoc.allocator.impl.rest.DbAllocatorResourceImpl$Proxy$_$$_WeldClientProxy.allocate(DbAllocatorResourceImpl$Proxy$_$$_WeldClientProxy.java)
[ERROR] 	at sun.reflect.GeneratedMethodAccessor115.invoke(Unknown Source)
[ERROR] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR] 	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:168)
[ERROR] 	at org.jboss.resteasy.core.ResourceMethod.invokeOnTarget(ResourceMethod.java:269)
[ERROR] 	at org.jboss.resteasy.core.ResourceMethod.invoke(ResourceMethod.java:227)
[ERROR] 	at org.jboss.resteasy.core.ResourceMethod.invoke(ResourceMethod.java:216)
[ERROR] 	at org.jboss.resteasy.core.SynchronousDispatcher.getResponse(SynchronousDispatcher.java:583)
[ERROR] 	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:565)
[ERROR] 	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:130)
[ERROR] 	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:208)
[ERROR] 	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:55)
[ERROR] 	at org.jboss.resteasy.plugins.server.servlet.HttpServletDispatcher.service(HttpServletDispatcher.java:50)
[ERROR] 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:847)
[ERROR] 	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:295)
[ERROR] 	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:214)
[ERROR] 	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:231)
[ERROR] 	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:149)
[ERROR] 	at org.jboss.as.web.security.SubjectInfoSetupValve.invoke(SubjectInfoSetupValve.java:34)
[ERROR] 	at org.jboss.as.web.security.SecurityContextAssociationValve.invoke(SecurityContextAssociationValve.java:169)
[ERROR] 	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:151)
[ERROR] 	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:97)
[ERROR] 	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:102)
[ERROR] 	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
[ERROR] 	at org.apache.coyote.http11.Http11Processor.process(Http11Processor.java:856)
[ERROR] 	at org.apache.coyote.http11.Http11Protocol$Http11ConnectionHandler.process(Http11Protocol.java:656)
[ERROR] 	at org.apache.tomcat.util.net.JIoEndpoint$Worker.run(JIoEndpoint.java:926)
[ERROR] 	at java.lang.Thread.run(Thread.java:748)
[ERROR] </pre></td>
[ERROR]       </tr>
[ERROR]     </table>
[ERROR]     <br/><br/>
[ERROR]     <div style="text-align:center">
[ERROR]       <a href='/Allocator'>Back to DB Allocator summary</a>
[ERROR]     </div>
[ERROR]   </body>
[ERROR] </html>)): DBAllocatorUnavailableException
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

```

A successful DB Release:

```
mvn org.keycloak.testsuite:db-allocator-plugin:6.0.0-SNAPSHOT:release -Ddballocator.uri="http://dballocator.keycloak.org:8080/Allocator/AllocatorServlet" -Ddballocator.properties.uuid="7a55eeeb-162e-465c-83ec-cac61143a09f"

[INFO] -------------< org.keycloak.testsuite:db-allocator-plugin >-------------
[INFO] Building DB Allocator Plugin 6.0.0-SNAPSHOT
[INFO] ----------------------------[ maven-plugin ]----------------------------
[INFO] 
[INFO] --- db-allocator-plugin:6.0.0-SNAPSHOT:release (default-cli) @ db-allocator-plugin ---
[info] Released database:
[info] -- UUID: 7a55eeeb-162e-465c-83ec-cac61143a09f
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.720 s
[INFO] Finished at: 2019-03-19T12:43:52+01:00
[INFO] ------------------------------------------------------------------------

```

Still to be done:
* [x] Create profile for the new plugin
* [x] Do all the wiring in the poms
* [x] Rewrite tests to Mockito
* [ ] Once the PR is green, create JIRAs for integrating the rest of DBs (into the pipeline).

slaskawi committed:
KEYCLOAK-9536 DB Allocator Plugin

hmlnarik commented on testsuite/integration-arquillian/db-allocator-plugin/pom.xml:
```diff
@@ -0,0 +1,86 @@
+        <dependency>
+            <groupId>org.keycloak.testsuite</groupId>
+            <artifactId>integration-arquillian-servers-app-server-undertow</artifactId>
+            <version>6.0.0-SNAPSHOT</version>
```
The version should be inherited from parent dependencyManagement

slaskawi replied:
Fixed. Thanks!

slaskawi replied:
hmmm this one is much more tricky than I thought!

It seems I introduced a cycle reference through `integration-arquillian-servers-app-server-spi`. I knew Maven will sort Undertor App Server correctly, but I haven't though about one layer below. 

This means I'll need to rework the testsuite for this module. I'll probably use Mockito to avoid similar "surprise"...

hmlnarik replied:
I believe test is not strictly needed here, so don't spend much time on rewriting it.

hmlnarik commented on testsuite/integration-arquillian/db-allocator-plugin/src/main/java/org/keycloak/testsuite/dballocator/AllocateDBMojo.java:
```diff
@@ -0,0 +1,145 @@
+    private String propertyPassword;
+
+    /**
+     * A property set as an output of this Mojo for DB Username.
```
```suggestion
     * A property set as an output of this Mojo for DB JDBC URL.
```

slaskawi replied:
Fixed. Thanks!

hmlnarik commented on testsuite/integration-arquillian/db-allocator-plugin/src/main/java/org/keycloak/testsuite/dballocator/AllocateDBMojo.java:
```diff
@@ -0,0 +1,145 @@
+    private String propertyDriver;
+
+    /**
+     * A property set as an output of this Mojo for DB Username.
```
needs correction

slaskawi replied:
Fixed. Thanks!

hmlnarik commented on testsuite/integration-arquillian/db-allocator-plugin/src/main/java/org/keycloak/testsuite/dballocator/ReleaseDBMojo.java:
```diff
@@ -0,0 +1,81 @@
+    private int totalRetries;
+
+    /**
+     * The number of retries for reaching the DB Allocator Service
```
needs correction

slaskawi replied:
Fixed. Thanks!

hmlnarik commented on testsuite/integration-arquillian/db-allocator-plugin/src/main/java/org/keycloak/testsuite/dballocator/AllocateDBMojo.java:
```diff
@@ -0,0 +1,145 @@
+    private String propertyUser;
+
+    /**
+     * A property set as an output of this Mojo for DB Username.
```
needs correction

slaskawi replied:
Fixed. Thanks!

hmlnarik reviewed (commented):
Thanks @slaskawi, nice one!

I think there should be support for `skip` property as well, and few JavaDoc nits, but otherwise looks very good!

vramik commented on testsuite/integration-arquillian/db-allocator-plugin/pom.xml:
```diff
@@ -0,0 +1,86 @@
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-plugin-plugin</artifactId>
+                <version>3.6.0</version>
```
Couldn't we use ${maven.version} here?

slaskawi replied:
I don't think this is a good idea to be honest. If I do that, the plugin compilation result might depend on your environment. 

Note, that Maven (at least the bits I used in the plugin) are backwards compatible. The plugin is compiled against `3.6.0` and I'm using `3.5.4` on my local box. Everything works fine with this setup.

slaskawi replied:
@vramik Could you please revisit this comment and tell me what you think?

vramik replied:
@slaskawi I've noticed there you've the property specified https://github.com/keycloak/keycloak/pull/5951/files#diff-b6a614e004ec399dc0a218c01b72965cR34 so my thinking was to use it. If there is a reason (and you described that it is) to remain current state, ignore my comment.

slaskawi replied:
Ok, thanks!

slaskawi replied:
Unfortunately it seems my approach was wrong. Since I'm using a more recent of Maven Core dependency, Maven freaked out when invoked on the prod pipeline. It seems using `${maven.version}` as you adviced might be a better choice (although we might run into some problems in the future, but I hope not :imp: ).

hmlnarik commented:
I've just noticed "Rewrite tests to Mockito" item in the checklist - mock libraries are banned from testsuite, so please exclude this item

slaskawi commented:
> I've just noticed "Rewrite tests to Mockito" item in the checklist - mock libraries are banned from testsuite, so please exclude this item

I can rewrite those tests without Mockito, but they will be twice as long (mostly because testing JAX-RS `Respose` objects is a bit painful, they contain a ton of abstract methods). 

The second argument is that Spring Security adapter already uses Mockito very intensively (https://github.com/keycloak/keycloak/blob/master/adapters/oidc/spring-security/pom.xml#L120). So I didn't add anything new to the testsuite, I just reused what we already had.

slaskawi commented on testsuite/integration-arquillian/servers/auth-server/jboss/pom.xml:
```diff
@@ -416,16 +416,19 @@
-                                                <property name="keycloak.connectionsJpa.user">${keycloak.connectionsJpa.user}</property>
-                                                <property name="keycloak.connectionsJpa.password">${keycloak.connectionsJpa.password}</property>
-                                                <property name="keycloak.connectionsJpa.schema">${keycloak.connectionsJpa.schema}</property>
+<!--                                                <property name="jdbc.driver.tmp.dir" value="${jdbc.driver.tmp.dir}" />-->
```
@vramik @tkyjovsk Here goes a very tricky part and I would like to know your opinion about it. However, before diving into my explanation, please read these two articles:

* https://technotes.khitrenovich.com/properties-resolution-maven-implications-antrun-plugin/
* https://books.sonatype.com/mvnref-book/reference/resource-filtering-sect-properties.html

The comment here suggests that you have been struggling with property resolution here. I also encountered some difficulties, because Maven resolves expressions when rendering the Effective POM. Since the DB Allocator Plugin manipulates `keycloak.connectionsJpa.*` properties in the runtime, resolving them for the Ant Plugin happened too quickly. This resulted in picking up properties defined [here](https://github.com/keycloak/keycloak/blob/31b767567e2c2940766c054a3298b73cfa8a7ec5/testsuite/integration-arquillian/pom.xml#L91-L96), instead of those changed by the DB Allocator Plugin.

The solution I propose is to force the Ant Plugin to resolve the properties instead of relying on the mechanism provided by Maven. Here's an example:

Maven resolution example:
```xml
<property name="jdbc.driver.tmp.dir">${jdbc.driver.tmp.dir}</property>
```

Ant Plugin resolution example:
```xml
<property name="jdbc.driver.tmp.dir" value="${jdbc.driver.tmp.dir}" />
```

Note, that the plugin now receives ''${jdbc.driver.tmp.dir}" and needs to resolve this based on the properties passed by Maven. So this time, the resolution happens in the runtime instead of rendering an Effective POM.

I'm not entirely sure if I didn't break anything with this change. So if you guys are invoking something manually somewhere, please check if it still works.

vramik replied:
If I understand it correctly the mechanisms you're proposing is already used e.g. [here](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/servers/auth-server/jboss/pom.xml#L219-L229)

slaskawi replied:
Yes, as you can see, this Pull Request replaces the piece you mentioned in your comment. In other words, instead of using:

```xml
<property name="jdbc.driver.tmp.dir">${jdbc.driver.tmp.dir}</property>
```

We will be using:

```xml
<property name="jdbc.driver.tmp.dir" value="${jdbc.driver.tmp.dir}" />
```

But as I said, I potentially might have broken something (please keep in mind that I'm in the team for quite a short time). So if you're using this properties somewhere (e.g. in prod testing or whatever), could you please run it with this patch to see if everything is still fine?

vramik replied:
If the overriding of the property from the command line still works: `mvn clean install -Djdbc.driver.tmp.dir=/my/favourite/folder/` it's fine to do the change you're proposing.

slaskawi replied:
I've just checked it and it seems to work fine.

slaskawi commented:
@keycloak-ci-bot test

hmlnarik commented:
@slaskawi There were several threads about using mock libraries in the past and the result is always the same - just do not use it. Feel free to search the keycloak-dev for references. Sadly, that means that this PR can go as it is.

I understand the benefits for testing this plugin though. So the viable path would be similar to spring-adapter - mockito needs to be be declared inside this project, and must go away from the top level pom.xml (including the version definition).

slaskawi commented on testsuite/db-allocator-plugin/pom.xml:
```diff
@@ -0,0 +1,82 @@
+        </dependency>
+        <dependency>
+            <groupId>org.apache.maven</groupId>
+            <artifactId>maven-core</artifactId>
```
This dependency might be redundant, see https://maven.apache.org/guides/plugin/guide-java-plugin-development.html

slaskawi replied:
Scope provided seems to work fine.

hmlnarik commented on testsuite/integration-arquillian/pom.xml:
```diff
@@ -363,7 +398,7 @@
         </profile>
         <profile>
-            <id>db-mysql</id>
+            <id>db-mysql57</id>
```
https://github.com/keycloak/keycloak/blob/master/docs/tests-db.md#using-built-in-profiles-to-run-database-tests-using-docker-containers needs to be updated as well

slaskawi replied:
Fixed, thanks!

hmlnarik commented on testsuite/integration-arquillian/HOW-TO-RUN.md:
```diff
@@ -717,6 +717,37 @@ The exact steps to configure Docker depend on the operating system.
+* `db-allocator-db-mssql2016` - for testing with MSSQL 2016
+* `db-allocator-db-mariadb` - for testing with MariaDB Galera Cluster
+* `db-allocator-db-postgres` - dor testing with Postgres 9.6.x
+* `db-allocator-db-mysql` - dor testing with MySQL 5.7
```
db-allocator-db-mysql57

slaskawi replied:
Fixed, thanks!

hmlnarik commented on testsuite/integration-arquillian/HOW-TO-RUN.md:
```diff
@@ -717,6 +717,37 @@ The exact steps to configure Docker depend on the operating system.
 The exact command line arguments depend on the operating system.
 
+
+## Using DB Allocator Service
```
This section belongs to https://github.com/keycloak/keycloak/blob/master/docs/tests-db.md

slaskawi replied:
Moved. Thanks!

hmlnarik commented on testsuite/integration-arquillian/HOW-TO-RUN.md:
```diff
@@ -717,6 +717,37 @@ The exact steps to configure Docker depend on the operating system.
+Using `-Dmaven.test.failure.ignore=true` is not strictly required but highly recommended. After running the tests,
+the DB Allocator Plugin should release the allocated database.
+
+Other implemented profiles:
```
Since this will be changing over time, I suggest to put one or two profiles here and refer user to the output of the following command to get the list of available db profiles:

```bash
mvn help:all-profiles -pl testsuite/integration-arquillian | grep -- db-
```

slaskawi replied:
hmmmm it would look inconsistent with the rest of the writing in `test-db.md`. Let me leave it like it is now.

slaskawi replied:
Fixed, thanks!

hmlnarik reviewed (commented):
Few doc comments.

hmlnarik reviewed (changes requested):
I like the plugin!

Few things to polish now that the functionally is correctly there (tried and tested):

```bash
mvn org.keycloak:db-allocator-plugin:allocate -e
[...]
[ERROR] Failed to execute goal org.keycloak:db-allocator-plugin:6.0.0-SNAPSHOT:allocate (default-cli) on project db-allocator-plugin: Execution default-cli of goal org.keycloak:db-allocator-plugin:6.0.0-SNAPSHOT:allocate failed.: NullPointerException -> [Help 1]
[...]
Caused by: java.lang.NullPointerException
    at java.net.URI$Parser.parse (URI.java:3042)
    at java.net.URI.<init> (URI.java:588)
    at java.net.URI.create (URI.java:850)
    at org.keycloak.testsuite.dballocator.AllocateDBMojo.execute (AllocateDBMojo.java:127)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
```

(same for `mvn org.keycloak:db-allocator-plugin:allocate -DpropertyDatabase=keycloak.connectionsJpa.database -DpropertyDriver=keycloak.connectionsJpa.driver -DpropertyURL=keycloak.connectionsJpa.url -DpropertyUser=keycloak.connectionsJpa.user -DpropertyPassword=keycloak.connectionsJpa.password`)

The default user should be obtained from `user.name` system property if not set on command line via `dballocator.user` property for `allocate` mojo.

I have also unresolved the docs since I don't understand the argument of being inconsistent with the rest of documentation. Could you elaborate? Please note that we need dynamic way for obtaining profiles also going forward for other types of resources (like LDAP servers) to test with and this would be the first instance of such dynamic listing among the docs.

slaskawi commented:
> I like the plugin!

@hmlnarik Thanks!

> `mvn org.keycloak:db-allocator-plugin:allocate -e` faliure

@hmlnarik This one has just been fixed. Thanks for spotting it!

> The default user should be obtained from user.name system property if not set on command line via dballocator.user property for allocate mojo.

@hmlnarik Done. 

> I have also unresolved the docs since I don't understand the argument of being inconsistent with the rest of documentation. Could you elaborate? Please note that we need dynamic way for obtaining profiles also going forward for other types of resources (like LDAP servers) to test with and this would be the first instance of such dynamic listing among the docs.

@hmlnarik As I clarified on the chat, I wanted to say, that the rest of the docs mentions all the profiles one by one. I implemented your suggestion and added mvn+grep trick to the document.

hmlnarik commented:
@slaskawi test
@keycloak-ci-bot test

hmlnarik commented:
@keycloak-ci-bot test

hmlnarik commented:
@keycloak-ci-bot test

slaskawi commented:
@keycloak-ci-bot test

hmlnarik commented:
@keycloak-ci-bot test

hmlnarik commented:
@keycloak-ci-bot test
